Add blog post: Using Finalizers to Control Deletion (#26997)
kubernetes/website: content/en/blog/_posts/2021-03-03-using-finalizers-to-control-deletion.md

## The basic `delete`

Kubernetes has several different commands you can use that allow you to create, read, update, and delete objects. For the purpose of this blog post, we’ll focus on four `kubectl` commands: `create`, `get`, `patch`, and `delete`. 

Here are examples of the basic `kubectl delete` command:

```
$ kubectl create configmap mymap
configmap/mymap created
```

```
$ kubectl get configmap/mymap
NAME    DATA   AGE
mymap   0      12s
```

```
$ kubectl delete configmap/mymap
configmap "mymap" deleted
```

```
$ kubectl get configmap/mymap
Error from server (NotFound): configmaps "mymap" not found
```

Shell commands preceded by `$` are followed by their output. You can see that we begin with a `kubectl create configmap mymap`, which will create the empty configmap `mymap`. Next, we need to `get` the configmap to prove it exists. We can then delete that configmap. Attempting to `get` it again produces an HTTP 404 error, which means the configmap is not found.

The state diagram for the basic `delete` command is very simple:

> ![State diagram for delete](/images/blog/2021-03-03-use-finalizers-to-control-deletion/state-diagram-delete.png)

Reviewer: This doesn't render - see https://deploy-preview-26997--kubernetes-io-master-staging.netlify.app/blog/2021/03/03/using-finalizers-to-control-deletion/

Author: interesting ... I'll reproduce and fix here.

  "apiVersion": "v1",
  "metadata": {
    "name": "test"
  },
  "spec": {
    "finalizers": null,
  }
}
EOF
```

This should be done with caution as it may delete the namespace only and leave orphan objects within the, now non-exiting, namespace - a confusing state for Kubernetes. If this happens, the namespace can be re-created manually and sometimes the orphaned objects will re-appear under the just-created namespace which will allow manual cleanup and recovery.

## Key Takeaways

As these examples demonstrate, finalizers can get in the way of deleting resources in Kubernetes, especially when there are parent-child relationships between objects. Often, there is a reason for adding a finalizer into the code, so you should always investigate before manually deleting it. Owner references allow you to specify and remove trees of resources, although finalizers will be honored in the process. Finally, the propagation policy can be used to specify the order of deletion via a custom API call, giving you control over how objects are deleted. Now that you know a little more about how deletions work in Kubernetes, we recommend you try it out on your own, using a test cluster. 

> All of the scenarios covered in this post are demonstrated in this video from a recent webinar: {{< youtube id="F7-ZxWwf4sY" title="Clean Up Your Room! What Does It Mean to Delete Something in K8s">}}

Reviewer: For better sizing, try:
```suggestion
{{< youtube youtube-quote-sm id="F7-ZxWwf4sY" title="Clean Up Your Room! What Does It Mean to Delete Something in K8s">}}
```

---
layout: blog
title: 'Using Finalizers to Control Deletion'

> date: 2021-03-03

Reviewer: Can I pencil in a future date here?
```suggestion
date: 2021-04-28
```

Author: 2021-04-28 works.

Author: I used your suggested date.
